components/webext-storage/src/store.rs

```diff
@@ -9,55 +9,93 @@ use std::path::Path;
     #[cfg(test)]
     pub fn new_memory(db_path: &str) -> Result<Self> {
         Ok(Self {
             db: StorageDb::new_memory(db_path)?,
         })
     }
 
-    // The "public API".
+    /// Sets one or more JSON key-value pairs for an extension ID. Returns a
+    /// list of changes, with existing and new values for each key in `val`.
     pub fn set(&self, ext_id: &str, val: JsonValue) -> Result<StorageChanges> {
-        let mut conn = self.db.writer.lock().unwrap();
-        let tx = conn.transaction()?;
+        let tx = self.db.unchecked_transaction()?;
         let result = api::set(&tx, ext_id, val)?;
         tx.commit()?;
         Ok(result)
     }
 
+    /// Returns the values for one or more keys `keys` can be:
+    ///
+    /// - `null`, in which case all key-value pairs for the extension are
+    ///   returned, or an empty object if the extension doesn't have any
+    ///   stored data.
+    /// - A single string key, in which case an object with only that key
+    ///   and its value is returned, or an empty object if the key doesn't
+    //    exist.
+    /// - An array of string keys, in which case an object with only those
+    ///   keys and their values is returned. Any keys that don't exist will be
+    ///   omitted.
+    /// - An object where the property names are keys, and each value is the
+    ///   default value to return if the key doesn't exist.
+    ///
+    /// This method always returns an object (that is, a
+    /// `serde_json::Value::Object`).
     pub fn get(&self, ext_id: &str, keys: JsonValue) -> Result<JsonValue> {
         // Don't care about transactions here.
-        let conn = self.db.writer.lock().unwrap();
-        api::get(&conn, ext_id, keys)
+        api::get(&self.db, ext_id, keys)
     }
 
+    /// Deletes the values for one or more keys. As with `get`, `keys` can be
+    /// either a single string key, or an array of string keys. Returns a list
+    /// of changes, where each change contains the old value for each deleted
+    /// key.
     pub fn remove(&self, ext_id: &str, keys: JsonValue) -> Result<StorageChanges> {
-        let mut conn = self.db.writer.lock().unwrap();
-        let tx = conn.transaction()?;
+        let tx = self.db.unchecked_transaction()?;
         let result = api::remove(&tx, ext_id, keys)?;
         tx.commit()?;
         Ok(result)
     }
 
+    /// Deletes all key-value pairs for the extension. As with `remove`, returns
+    /// a list of changes, where each change contains the old value for each
+    /// deleted key.
     pub fn clear(&self, ext_id: &str) -> Result<StorageChanges> {
-        let mut conn = self.db.writer.lock().unwrap();
-        let tx = conn.transaction()?;
+        let tx = self.db.unchecked_transaction()?;
         let result = api::clear(&tx, ext_id)?;
         tx.commit()?;
         Ok(result)
     }
+
+    /// Tears down the store, closing its underlying database connection.
+    pub fn teardown(self) -> Result<()> {
```

Comment:
If you're wondering where this will get called... 😊 [Here!](https://hg.mozilla.org/mozilla-central/file/tip/toolkit/components/extensions/storage/webext_storage_bridge/src/store.rs#l50)